modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/IndexBuilderTxRwOperationTracker.java

```diff
@@ -0,0 +1,203 @@
+            CompletableFuture<Void> awaitFuture = awaitCompleteOperationsFutureByCatalogVersion.computeIfAbsent(
+                    catalogVersion,
+                    i -> new CompletableFuture<>()
+            );
+
+            if (catalogVersion <= minAllowedCatalogVersionForStartOperation.get()) {
+                awaitCompleteOperationsFutureByCatalogVersion.remove(catalogVersion);
+
+                awaitFuture.complete(null);
+            }
+
+            // Let's prevent freezing.
+            awaitFuture = orCloseFuture(awaitFuture);
+
+            return awaitFuture.thenCompose(unused -> {
+                // Let's select counters for versions up to (not including) the requested catalog version.
+                CompletableFuture<?>[] futures = operationCounterByCatalogVersion.headMap(catalogVersion, false).values().stream()
+                        .map(TxRwOperationCounter::operationsFuture)
+                        .toArray(CompletableFuture[]::new);
+
+                // Let's prevent freezing.
+                return orCloseFuture(allOf(futures));
+            });
+        });
+    }
+
+    /**
+     * Increases the count of transaction operations by {@code 1}, at the start of a new RW transaction operation.
+     *
+     * <p>It is expected that after the operation is completed, {@link #decrementOperationCount(int)} will be called.</p>
+     *
+     * @param catalogVersion Catalog version at the beginning of the RW transaction of operation.
+     * @throws RejectTxRwOperationException If the passed catalog version is less than the minimum available to start a RW transaction
+     *      operation.
+     */
+    void incrementOperationCount(int catalogVersion) {
+        inBusyLock(busyLock, () -> {
+            operationCounterByCatalogVersion.compute(catalogVersion, (i, txRwOperationCounter) -> {
+                if (txRwOperationCounter == null) {
+                    return new TxRwOperationCounter();
+                }
+
+                return txRwOperationCounter.incrementOperationCount();
+            });
+
+            if (catalogVersion < minAllowedCatalogVersionForStartOperation.get()) {
+                decrementOperationCount(catalogVersion);
+
+                throw new RejectTxRwOperationException();
+            }
+        });
+    }
+
+    /**
+     * Decreases the count of transaction operations by {@code 1}, at the finish of a new RW transaction operation.
+     *
+     * <p>It is expected that {@link #incrementOperationCount(int)} was previously called.</p>
+     *
+     * @param catalogVersion Catalog version at the beginning of the RW transaction of operation.
+     */
+    void decrementOperationCount(int catalogVersion) {
+        inBusyLock(busyLock, () -> {
+            // I intentionally create a local holder so as not to complete the TxRwOperationCounter#operationsFuture inside Map#compute.
+            TxRwOperationCounter[] localHolder = new TxRwOperationCounter[1];
```

Comment:
I think that the name is not suitable, I will leave it as it is.

Reply:
`localHolder` does not tell you *why* this variable is needed. It would be cool if it did. The name that I suggested attempts to provide this info. If it's not good enough (by the way, why? You gave no explanation), let's invent a better name.

Reply:
The name you suggested in this situation does not bring me any improvement; on the contrary, it confuses me. The name is short enough and there is not a lot of code to make it difficult to understand.

The current dialogue is like a matter of taste.